examples/logreg/tfe-logreg.py

```diff
@@ -43,27 +34,31 @@ def consume(v: Union[np.ndarray, tf.Tensor], prot: tfe.protocol.Pond) -> PondPri
 
     W = prot.define_private_variable(tf.random_uniform([nb_feats, 1], -0.01, 0.01))
     b = prot.define_private_variable(tf.zeros([1]))
-    xp = consume(x, prot)  # We secure our inputs
-    yp = consume(y, prot)
+    xp = prot.define_private_input(x)  # We secure our inputs
```

Comment:
not convinced by this tbh, we're opening up for security issues (even if a warning is printed) and in this case it doesn't seem needed since both `x` and `y` are the output of a function (see my suggestion in https://github.com/mortendahl/tf-encrypted/issues/149)

Reply:
I’ve been thinking about the current problem for `define_private_input`, I came to the conclusion that the name itself is misleading. `define_private_input` is not really defining anything, it secures a provider for now. `prot.secure_provider` would be a better name 

I think one of the most complicated thing to understand (from a new user point of view) is that you want "the secure step" to happens on the machine that provide the input but you can define the code that provide the input on the machine running the main script.
We achieve that by creating a function that load the inputs and use TF to execute the function on a different  machine, in the meantine we sneak in out secure step by wrapping the function outputs. All of that led to the notion of InputProvider probably.

I would argue that it would be better to have directly:
```python
def secure_input_function(func, player=default_input_player):
...
```
I don't like to have objects for everything, the more we can have "pure function" the better we are.

Side note: I would agree with you that this can be a decoration.

What do you think?

Reply:
renaming to something like `private_input` would make sense to me, but `secure_provider` less so ("private input"/"public input" is somewhat normal in MPC literature).

letting the input method on `prot` take either objects or functions with players could both work, but I think decorators is actually better as it not only allows us to hide this stuff, it also gathers the function definition and where it's executed, as well as make it easy to use default prot objects:
```python
@tfe.private_input('prediction-client')
def get_input():
    return tf.constant([4,3,2,1])

x = get_input()
```
instead of
```python
def get_input():
    return tf.constant([4,3,2,1])

with ... as prot:
   x = prot.private_input(get_input, "prediction-client")
```

Reply:
Interesting.
From my engineer point of view I would rather see:
```python
@tfe.run_on('prediction-client')
def get_input():
    return tf.constant([4,3,2,1])
```
or maybe
```python
@tfe.secure_on('prediction-client')
def get_input():
    return tf.constant([4,3,2,1])
```

Reply:
yeah I see what you mean. you need to specify if you want it as private or public values though (`PondPrivateTensor`/`PondPublicTensor`), at least I think it would be useful to be able to have both

Reply:
any thought @justin1121 @bendecoste @jvmancuso @ianlivingstone ?

Reply:
I think we need to be able to define both public/private inputs this way. So I think I like `@tfe.private_input('prediction-client')` the best. 

Although, we could also pass a public/private flag into the `secure_on` function. Like `@tfe.secure_on('prediction-client', private=True)`. But, then the `secure_on` name doesn't make sense if it's public. Maybe `provide` would be better.

Think I'm on the side of `private_input`, in the end. Its an interesting balance between what would make sense for the average user and what would make sense to someone who knows a lot about MPC.